File: 2025/10/main_slow.zig
const std = @import("std");
const mem = std.mem;
const Allocator = mem.Allocator;
const ArenaAllocator = std.heap.ArenaAllocator;
const ArrayList = std.ArrayList;

const Button = []const usize;
const Lights = []bool;

const Machine = struct {
    goal: Lights,
    buttons: []Button,
    backing_allocator: Allocator,
    arena: ArenaAllocator,

    fn init(backing: Allocator, line: []const u8) !Machine {
        var arena = std.heap.ArenaAllocator.init(backing);
        const gpa = arena.allocator();
        var goal: ArrayList(bool) = .empty;
        {
            const start = mem.indexOfScalar(u8, line, '[').?;
            const end = mem.indexOfScalar(u8, line, ']').?;
            try goal.ensureTotalCapacity(gpa, end - start);
            for (line[start + 1 .. end]) |c| {
                switch (c) {
                    '.' => goal.appendAssumeCapacity(false),
                    '#' => goal.appendAssumeCapacity(true),
                    else => return error.InvalidInput,
                }
            }
        }

        var buttons: std.ArrayList(Button) = .empty;
        {
            var end: usize = 0;
            var start: usize = 0;
            while (true) {
                start = mem.indexOfScalarPos(u8, line, end + 1, '(') orelse break;
                end = mem.indexOfScalarPos(u8, line, start, ')').?;
                var it = mem.splitScalar(u8, line[start + 1 .. end], ',');
                var button: ArrayList(usize) = .empty;
                while (it.next()) |bs| {
                    try button.append(gpa, try std.fmt.parseUnsigned(usize, bs, 10));
                }
                try buttons.append(gpa, try button.toOwnedSlice(gpa));
            }
        }

        return .{
            .goal = try goal.toOwnedSlice(gpa),
            .buttons = try buttons.toOwnedSlice(gpa),
            .backing_allocator = backing,
            .arena = arena,
        };
    }

    fn deinit(self: *Machine) void {
        self.arena.deinit();
    }
};

const State = struct {
    prev: ?*State = null,
    current: Lights,
    next: ?[]State = null,
    backing_allocator: Allocator,
    arena: ArenaAllocator,

    fn init(backing: Allocator, current: Lights, prev: ?*State) State {
        return .{
            .backing_allocator = backing,
            .arena = ArenaAllocator.init(backing),
            .current = current,
            .prev = prev,
        };
    }

    fn initRoot(backing: Allocator, light_size: usize) !State {
        var arena = ArenaAllocator.init(backing);
        const gpa = arena.allocator();
        const current = try gpa.alloc(bool, light_size);
        return .{
            .backing_allocator = backing,
            .arena = arena,
            .current = current,
            .prev = null,
        };
    }

    fn deinit(self: *State) void {
        self.arena.deinit();
    }

    fn populate_next(self: *State, buttons: []Button) !void {
        const gpa = self.arena.allocator();
        if (self.next) |_| return;
        var next: std.ArrayList(State) = .empty;
        try next.ensureTotalCapacity(gpa, buttons.len);

        for (buttons) |button| {
            var new_mstate = try gpa.dupe(bool, self.current);
            for (button) |i| {
                new_mstate[i] = !new_mstate[i];
            }
            next.appendAssumeCapacity(State.init(self.backing_allocator, new_mstate, self));
        }

        self.next = try next.toOwnedSlice(gpa);
    }
};

fn bfs_state(root: *State, m: *Machine) !u64 {
    var q: [1 << 10]*State = undefined;
    var head: usize = 0;
    var tail: usize = 0;
    q[tail] = root;
    tail += 1;

    while (head != tail) {
        const state: *State = q[head];
        head += 1;
        if (mem.eql(bool, state.current, m.goal)) {
            var dist: usize = 0;
            var cur = state;
            while (cur.prev) |prev| {
                cur = prev;
                dist += 1;
            }
            return dist;
        }

        try state.populate_next(m.buttons);
        for (state.next.?) |*branch| {
            q[tail] = branch;
            tail += 1;
        }
    }

    std.debug.print("bfs search queue is empty or full\n", .{});
    std.process.abort();
}

pub fn main() !void {
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const alloc = arena.allocator();

    const argv = try std.process.argsAlloc(alloc);
    if (argv.len < 2) {
        std.debug.print("missing input argv\n", .{});
        return;
    }
    const file = try std.fs.openFileAbsolute(argv[1], .{ .mode = .read_only });
    defer file.close();

    var buffer: [1024]u8 = undefined;
    var file_reader = file.reader(&buffer);
    const reader = &file_reader.interface;

    var sum: u64 = 0;
    while (reader.takeDelimiterInclusive('\n')) |line| {
        var m = try Machine.init(alloc, line);
        var root = try State.initRoot(alloc, m.goal.len);
        sum += try bfs_state(&root, &m);
    } else |err| {
        if (err != error.EndOfStream) {
            std.debug.print("{}", .{err});
        }
    }
    std.debug.print("{d}\n", .{sum});
}

File: 2025/10/main_slow.cpp
#include "../common/common.h"
#include <cstdio>
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <string>

using button = vector<usize>;

struct machine {
  vector<bool> goal;
  vector<button> buttons;

  friend std::ostream& operator<<(std::ostream& os, const machine& m) {
    os << "[";
    for (auto g : m.goal) { os << (g ? '#' : '.'); }
    os << "] ";

    for (const auto& b : m.buttons) {
      os << "(";
      auto b_size = b.size();
      for (auto i{0}; i < b_size; i++) {
        os << b[i];
        if (i != b_size - 1) os << ",";
      }
      os << ") ";
    }
    os << std::endl;
    return os;
  }
};

machine parse_machine(const string& line) {
  vector<bool> goal{};
  {
    auto start = line.find("[");
    auto end = line.find("]");
    string goal_string = line.substr(start + 1, end - start - 1);
    for (auto& c : goal_string) {
      if (c == '.') {
        goal.emplace_back(false);
      } else if (c == '#') {
        goal.emplace_back(true);
      } else {
        std::printf("invalid goal string: %s\n", line.c_str());
      }
    }
  }

  vector<button> buttons{};
  {
    auto end = 0;
    auto start = 0;
    while (true) {
      start = line.find("(", end + 1);
      end = line.find(")", start);
      if (start == string::npos) break;
      auto button_strings = split(line.substr(start + 1, end - start - 1), ",");
      button b{};
      for (const auto& bs : button_strings) { b.emplace_back(std::stoul(bs)); }
      buttons.emplace_back(b);
    }
  }
  return {goal, buttons};
}

struct state {
  state* prev = nullptr;
  vector<bool> current;
  vector<state> next = {};

  state(usize len) { current = vector<bool>(len, false); }
  state(vector<bool> v, state* prev = nullptr) : current(std::move(v)), prev(prev) {}

  void populate_next_states(const vector<button>& buttons) {
    if (!next.empty()) return;
    next.reserve(buttons.size());
    for (const auto& button : buttons) {
      auto new_current{current};
      for (auto light : button) { new_current[light] = !new_current[light]; }
      next.emplace_back(std::move(new_current), this);
    }
  }

  friend std::ostream& operator<<(std::ostream& os, const state& s) {
    os << "[";
    for (auto l : s.current) { os << (l ? '#' : '.'); }
    os << "] \n";

    for (const auto& n : s.next) { os << " - " << n << std::endl; }
    return os;
  }
};

u64 bfs_state(state& root, const machine& m) {
  std::deque<state*> q{};
  q.push_back(&root);

  while (!q.empty()) {
    state* node{q.front()};
    q.pop_front();
    if (node->current == m.goal) {
      usize dist{};
      for (state* cur = node; cur->prev != nullptr; cur = cur->prev) dist++;
      return dist;
    }

    node->populate_next_states(m.buttons);
    for (auto& branch : node->next) { q.push_back(&branch); }
  }

  std::printf("Something went wrong, bfs search queue is empty\n");
  std::abort();
}

int main(int argc, char* argv[]) {
  if (argc < 2) {
    std::cerr << "missing input argv" << std::endl;
    return 1;
  }
  std::ifstream input(argv[1]);
  if (!input) {
    std::cerr << "input file cannot be opened" << std::endl;
    return 1;
  }
  u64 sum{};
  string line{};
  while (std::getline(input, line)) {
    // do something
    machine m = parse_machine(line);
    state s(m.goal.size());
    sum += bfs_state(s, m);
  }
  std::cout << sum << std::endl;
  return 0;
};
